### PR TITLE
Build packages on non-Windows platforms

### PR DESCRIPTION
**Pull request by karajas**
/cc @ellismg

### CODE REVIEW

**Comment by karajas** [comment]
For context, we would like to be able to build and consume vstest packages on non-Windows distros as a part of the CLI build.

**Comment by karajas** [comment]
Hi @Faizan2304 , could you please take a look at this PR?

File: src/package/nuspec/TestPlatform.Build.nuspec
Comment (ellismg) [inline src/package/nuspec/TestPlatform.Build.nuspec]: Was this needed because the *nix build was putting artifacts in a different location?

File: scripts/build.sh
Comment (karajas) [inline scripts/build.sh]: Will fix.

File: src/package/nuspec/TestPlatform.Build.nuspec
Comment (karajas) [inline src/package/nuspec/TestPlatform.Build.nuspec]: Yeah on windows it goes to artifacts\Debug\Microsoft.TestPlatform.Build\netstandard1.3, on *nix it goes to artifacts/Debug/Microsoft.TestPlatform.Build/Debug/netstandard1.3/ . I can investigate why this happens instead.

Comment (codito) [inline src/package/nuspec/TestPlatform.Build.nuspec]: Can you please investigate?

**Comment by codito** [comment]
Are these changes required in `15.0-rtm` branch (so that RTM bits can be packaged)?
The binaries are default public signed. Do we require strongname signing? In that case, we may need to create a CI on devdiv.vs.com?

What is the validation for this - is it sufficient to ensure nuget packages are created? I am asking this because we need to ensure build script keeps working in case the csproj/nuspec are refactored. Probably we should add a Ubuntu jenkins ci.

File: scripts/build.sh
Comment (codito) [inline scripts/build.sh]: This is probably no-op since the net46 assemblies for testhost are not built.

File: src/package/nuspec/Microsoft.TestPlatform.Build.csproj
Comment (codito) [inline src/package/nuspec/Microsoft.TestPlatform.Build.csproj]: nit: replace tabs with spaces.

**Comment by karajas** [comment]
> Are these changes required in 15.0-rtm branch (so that RTM bits can be packaged)?

I think master is the branch we want to grab at the moment to get the latest bits to test in the context of the CLI. We may need 15.0-rtm in the future. @ellismg ?

> The binaries are default public signed. Do we require strongname signing? In that case, we may need to create a CI on devdiv.vs.com?

We probably don't need signing at the moment since we aren't going to publicly publish these packages on any feed. They are going to be used as a transport package between repos when we build the CLI.

> What is the validation for this - is it sufficient to ensure nuget packages are created? I am asking this because we need to ensure build script keeps working in case the csproj/nuspec are refactored. Probably we should add a Ubuntu jenkins ci.

Yup package creation is the requirement. We have a CI setup to ensure that these packages are still built and consumed correctly. But I agree that we should probably setup another CI for this - I'll follow up in an issue.

File: src/package/nuspec/Microsoft.TestPlatform.Build.csproj
Comment (karajas) [inline src/package/nuspec/Microsoft.TestPlatform.Build.csproj]: will update.

File: scripts/build.sh
Comment (karajas) [inline scripts/build.sh]: Oops agreed, will remove.

File: src/package/nuspec/TestPlatform.Build.nuspec
Comment (karajas) [inline src/package/nuspec/TestPlatform.Build.nuspec]: This was a wildcard filtering issue that should be resolved now.

File: src/package/nuspec/TestPlatform.Build.nuspec
Comment (karajas) [inline src/package/nuspec/TestPlatform.Build.nuspec]: Will this affect servicing any of the packages on official/dev feeds?

**Comment by ellismg** [comment]
>>Are these changes required in 15.0-rtm branch (so that RTM bits can be packaged)?
>I think master is the branch we want to grab at the moment to get the latest bits to test in the context of the CLI. We may need 15.0-rtm in the future. @ellismg ?

We need these changes for vNext of .NET Core CLI.  They don't need to be tied to VS 2017.  If you have a branch for Update 1 already, I think we'd want to be there.  Otherwise, whatever your dev branch is.

>>The binaries are default public signed. Do we require strongname signing? In that case, we may need to create a CI on devdiv.vs.com?
>We probably don't need signing at the moment since we aren't going to publicly publish these packages on any feed. They are going to be used as a transport package between repos when we build the CLI.

Having them public signed is great.  We don't need them to be real signed.  One of the goals of this effort is to enable folks outside of Microsoft to build the entire .NET Core product from source and in that world they *have* to public sign for assemblies that are signed with a key hasn't been open sourced.

>>What is the validation for this - is it sufficient to ensure nuget packages are created? I am asking this because we need to ensure build script keeps working in case the csproj/nuspec are refactored. Probably we should add a Ubuntu jenkins ci.
>Yup package creation is the requirement. We have a CI setup to ensure that these packages are still built and consumed correctly. But I agree that we should probably setup another CI for this - I'll follow up in an issue.

Our composition logic should also catch places where this regresses, but if you can get CI coverage in Jenkins, that would be great.

File: scripts/build.sh
Comment (codito) [inline scripts/build.sh]: Can we use `stagingDir` as output directory for packages?

File: scripts/build.sh
Comment (codito) [inline scripts/build.sh]: Does `Microsoft.NET.Test.Sdk.nuspec` need to be with `"` for consistency?

File: scripts/build.sh
Comment (karajas) [inline scripts/build.sh]: Sure I'll update this.

**Comment by karajas** [comment]
@codito Please take a look.

File: src/package/nuspec/TestPlatform.Build.nuspec
Comment (codito) [inline src/package/nuspec/TestPlatform.Build.nuspec]: yes, this will break official nuget packages. Why is this required? I tried to mimic the steps we're doing (create artifacts\Debug\MS.TP.Build\netstandard1.3 directory structure, copy MS.TP.Build.csproj to artifacts\Debug, then call dotnet pack; it seemed to package binaries without this change).

**Comment by codito** [comment]
Will setup a jenkins CI once we merge this.

File: src/package/nuspec/TestPlatform.Build.nuspec
Comment (karajas) [inline src/package/nuspec/TestPlatform.Build.nuspec]: @codito Actually this wouldn't break official nuget packages - it still gets dropped in lib\netstandard1.3 inside the folder.
I might be confused about https://github.com/Microsoft/vstest/pull/552/files#diff-56d541dcd7f193797f0c7bc290c82117L354, are the files not supposed to go in separate folders based on framework ( a netstandard one and a net46 one) ? Because it seemed as though those folders weren't created according to the comment above it. I'm happy to have the linux packages do the same, just wanted to double check first.

Comment (codito) [inline src/package/nuspec/TestPlatform.Build.nuspec]: The packaging may not depend on configuration, the nuspec will always expect Debug bits. In official CI, we build release configuration.

> I might be confused about https://github.com/Microsoft/vstest/pull/552/files#diff-56d541dcd7f193797f0c7bc290c82117L354, are the files not supposed to go in separate folders based on framework ( a netstandard1.3 one and a net46 one)

You're right, they should go in separate folders. Today only a directory is created for `netstandard1.3` is because of the typo in `$binariesDirectory = [System.IO.Path]::Combine("src", "$packageName", "bin", "$TPB_Configuration", "*.*")`; the pattern should be `*`, not `*.*`.

File: scripts/build.ps1
Comment (codito) [inline scripts/build.ps1]: As we discussed below, we should change the pattern to `*` instead of `*.*`.

File: src/package/nuspec/TestPlatform.Build.nuspec
Comment (karajas) [inline src/package/nuspec/TestPlatform.Build.nuspec]: Will update shortly.

File: scripts/build.ps1
Comment (karajas) [inline scripts/build.ps1]: see below

File: src/package/nuspec/TestPlatform.Build.nuspec
Comment (karajas) [inline src/package/nuspec/TestPlatform.Build.nuspec]: The only addition change I had to do was remove the backslash so that the *nix package does not create an extra folder.

**Comment by karajas** [comment]
Can I move the nupkgs to a separate artifacts folder?

**Review by codito — APPROVED** [review]

**Comment by codito** [comment]
> Can I move the nupkgs to a separate artifacts folder?

Please go ahead. Changes look good to me, will wait for this change and merge this PR.

**Comment by karajas** [comment]
This has a missing quote. Sorry will open up a separate pr. https://github.com/Microsoft/vstest/pull/552/files#diff-d99b26f864526f6f8f79861d49deb922R284

**Pull request merged**
